### PR TITLE
accelerate build table def when limit 0

### PR DESCRIPTION
**Pull request by YANGGMM**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #https://github.com/matrixorigin/MO-Cloud/issues/3122 #https://github.com/matrixorigin/MO-Cloud/issues/3120

## What this PR does / why we need it:
accelerate build table def when limit 0

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added functionality to accelerate table access when the SQL query has a limit of zero.
- Implemented methods to build table definitions from column metadata.
- Enhanced the query builder to skip plan building when the limit is zero, optimizing performance.
- Added new test cases and SQL scripts to verify the zero limit optimization.
- Updated protocol buffers to include a field for skipping plan building.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>authenticate.go</strong><dd><code>Add SQL format and function for table column definitions</code>&nbsp; </dd></summary>
<hr>

pkg/frontend/authenticate.go

<li>Added a new SQL format string for fetching table column definitions.<br> <li> Implemented a function to get SQL for table column definitions.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-849f201c351210bd95807e99d1538e2602a5244b256c35e58467afe304c509e6">+10/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>compiler_context.go</strong><dd><code>Add method to build table definitions from columns</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/frontend/compiler_context.go

- Added a method to build table definitions from columns.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-2e1e8cbf97f9cd34095a0a6b95d33456cb53dad2c6fec7eb9cd37893b8048b8b">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>util.go</strong><dd><code>Implement function to build table definitions from metadata</code></dd></summary>
<hr>

pkg/frontend/util.go

<li>Implemented function to build table definitions from column metadata.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-005617216374ac58bf3cf72b81841d5cb6a17495d1d7b36ff387e294a0d6e043">+103/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>sql_executor_context.go</strong><dd><code>Add method stub for building table definitions by columns</code></dd></summary>
<hr>

pkg/sql/compile/sql_executor_context.go

- Added a method stub for building table definitions by columns.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-ce4568890a0654c1226626ec5fe741ba7feed78c92b97bab59961d248808fa8c">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>query_builder.go</strong><dd><code>Enhance query builder for zero limit optimization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/sql/plan/query_builder.go

<li>Enhanced query builder to skip plan building when limit is zero.<br> <li> Added logic to handle table definition building when skipping plan.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-144e02a38da50867dc021b9254d10a5fd131671cca344c2323337ba2141440f1">+46/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>types.go</strong><dd><code>Add interface method for building table definitions by columns</code></dd></summary>
<hr>

pkg/sql/plan/types.go

- Added interface method for building table definitions by columns.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-fd44fd4ef13ca8bef0473460d6b4298a1beb5cd259562bd013e35433af2b80fb">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>compiler_context.go</strong><dd><code>Add method stub for building table definitions by columns</code></dd></summary>
<hr>

pkg/vm/engine/memoryengine/compiler_context.go

- Added method stub for building table definitions by columns.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-c2c5a10930b5c592fb5880315756ef1a7e7d58f4e3311de4da573bb3e87574ca">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>plan.proto</strong><dd><code>Add skip plan building field to Query message</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

proto/plan.proto

- Added a new field to the Query message to skip plan building.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-22bd2bd1e47d7acf42499abc93266a657dfd1a35f32837167470fd926564dfb5">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>mock.go</strong><dd><code>Add mock method for building table definitions by columns</code></dd></summary>
<hr>

pkg/sql/plan/mock.go

- Added mock method for building table definitions by columns.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-2c86cf181afbb13bcc985c55bceef3a4e2291d431ee4cd3eede001dc26d55980">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>types_mock.go</strong><dd><code>Add mock implementation for building table definitions by columns</code></dd></summary>
<hr>

pkg/sql/plan/types_mock.go

<li>Added mock implementation for building table definitions by columns.<br>

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-dd406f5de662b1388c36fcba1ead0a98351aad279c7c30905a31c91387f30dc5">+15/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>system.result</strong><dd><code>Add test cases for zero limit optimization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/distributed/cases/system/system.result

- Added test cases for zero limit optimization.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-4cc775ef882cf26f54fe270654819b2a4e1dc7950b9aa933c533c88c15907940">+117/-1</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>system.sql</strong><dd><code>Add SQL test scripts for zero limit optimization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/distributed/cases/system/system.sql

- Added SQL test scripts for zero limit optimization.

</details>

  </td>
  <td><a href="https://github.com/matrixorigin/matrixone/pull/18707/files#diff-a302bac409e6feba837280837bae0d359943f44c3b9c36334c7d37b55b88bda2">+102/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: pkg/sql/plan/query_builder.go [hunk 1]
```diff
@@ -2251,6 +2251,28 @@ func (builder *QueryBuilder) buildSelect(stmt *tree.Select, ctx *BindContext, is
...
+					limitBinder := NewLimitBinder(builder, ctx)
+					limitExpr, err := limitBinder.BindExpr(astLimit.Count, 0, true)
+					if err != nil {
+						return 0, err
+					}
+
+					if cExpr, ok := limitExpr.Expr.(*plan.Expr_Lit); ok {
+						if c, ok := cExpr.Lit.Value.(*plan.Literal_U64Val); ok {
+							if c.U64Val == 0 {
+								builder.qry.SkipBuildPlanTag = true
```
Comment (badboynt1) [inline pkg/sql/plan/query_builder.go:2268]: 把这个字段放在builder里面，而不是qry里面是不是更好？

**Review by daviszhen — APPROVED** [review]

**Review by badboynt1 — APPROVED** [review]

**Review by m-schen — APPROVED** [review]

**Review by heni02 — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]
